c-scape/src/threads/key.rs

```diff
@@ -0,0 +1,183 @@
+use alloc::boxed::Box;
+use core::cell::Cell;
+use core::ptr::{null, null_mut};
+use core::sync::atomic::{AtomicU16, Ordering};
+use errno::{set_errno, Errno};
+use libc::{c_int, c_void};
+
+use origin::sync::RwLock;
+
+const PTHREAD_KEYS_MAX: u32 = 1024;
+const PTHREAD_DESTRUCTOR_ITERATIONS: u8 = 4;
+
+#[derive(Clone, Copy)]
+struct KeyData {
+    next_key: libc::pthread_key_t,
+    destructors: [Option<unsafe extern "C" fn(_: *mut c_void)>; PTHREAD_KEYS_MAX as usize],
+}
+
+#[derive(Clone, Copy)]
+struct ValueWithEpoch {
+    epoch: u16,
+    data: *mut c_void,
+}
+
+impl ValueWithEpoch {
+    const fn new() -> Self {
+        ValueWithEpoch {
+            epoch: 0,
+            data: null_mut(),
+        }
+    }
+}
+
+static KEY_DATA: RwLock<KeyData> = RwLock::new(KeyData {
+    next_key: 0,
+    destructors: [None; PTHREAD_KEYS_MAX as usize],
+});
+
+static EPOCHS: [AtomicU16; PTHREAD_KEYS_MAX as usize] =
+    [const { AtomicU16::new(0) }; PTHREAD_KEYS_MAX as usize];
+
+// This uses an epoch-based system for differentiating between
+// reused keys corresponding to the same slot.
+#[thread_local]
+static VALUES: [Cell<ValueWithEpoch>; PTHREAD_KEYS_MAX as usize] =
+    [const { Cell::new(ValueWithEpoch::new()) }; PTHREAD_KEYS_MAX as usize];
+
+#[thread_local]
+static HAS_REGISTERED_CLEANUP: Cell<bool> = Cell::new(false);
+
+#[no_mangle]
+unsafe extern "C" fn pthread_getspecific(key: libc::pthread_key_t) -> *mut c_void {
+    libc!(libc::pthread_getspecific(key));
+
+    let latest_epoch = EPOCHS[key as usize].load(Ordering::SeqCst);
+    let ValueWithEpoch { epoch, data } = VALUES[key as usize].get();
+
+    // If the latest epoch is newer, then that means this slot got reallocated.
+    // Either we are reusing a deleted key, which is UB, or we are reusing
+    // the new key, which must return null initially.
+    if epoch < latest_epoch {
+        null_mut()
+    } else {
+        data
+    }
+}
+
+#[no_mangle]
+unsafe extern "C" fn pthread_setspecific(key: libc::pthread_key_t, value: *const c_void) -> c_int {
+    libc!(libc::pthread_setspecific(key, value));
+
+    // If this is the first-time we have gotten here,
+    // we need to actually register the dtors for cleanup.
+    if !HAS_REGISTERED_CLEANUP.get() {
+        // TODO: The ordering that we need here is
+        // all the pthread_cleanup functions,
+        // then all thread-local dtors,
+        // and then finally all static
+        // (`__cxa_thread_atexit_impl`) dtors.
```

Comment:
The first two are specified by POSIX:

> Any clean-up handlers established by [pthread_cleanup_push(3)](https://man7.org/linux/man-pages/man3/pthread_cleanup_push.3.html) that
 have not yet been popped, are popped (in the reverse of the order
 in which they were pushed) and executed.  If the thread has any
 thread-specific data, then, after the clean-up handlers have been
 executed, the corresponding destructor functions are called, in
 an unspecified order.

And the last two are specced by the C++ standard:

> Destructors for initialized objects with thread storage duration within a given thread are called as a result of returning from the initial function of that thread and as a result of that thread calling std::exit. The completions of the destructors for all initialized objects with thread storage duration within that thread are sequenced before the initiation of the destructors of any object with static storage duration.

Reply:
Looking at this, I noticed that origin wasn't calling the TLS dtors on exit. I've now filed https://github.com/sunfishcode/mustang/pull/133 for that.

Other than that, and other than `ptthread_cleanup_push`, I think we're ok.

Reply:
> other than `ptthread_cleanup_push`

I'm not quite sure what you're referring to here.

Reply:
I mean that we haven't implemented the ordering for destructors registered with `pthread_cleanup_push` yet, but that's ok for now because we haven't implemented `pthread_cleanup_push` yet.